### ISSUE TITLE
External IdP integration with ADFS may fail.

### ISSUE THREAD

**Issue by y-tabata** [keycloak/keycloak#16154]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

identity-brokering

### Describe the bug

After version up from 18.0.2 to 20.0.1, external IdP integration with ADFS may fail.

The condition is when the state parameter in an authorization request includes "%2B".
In this case, ADFS returns the state parameter in an authorization response that includes "+" instead of "%2B". 

Generally thinking, this is an ADFS bug and ADFS should encode "+" to "%2B".

However, ideally, Keycloak should not send a confusing state value that can be read in two ways like "+" and "%2B".

The cause is this PR.
https://github.com/keycloak/keycloak/pull/13757

ADFS's version
OS Name:                   Microsoft Windows Server 2019 Datacenter
OS Version:                10.0.17763 N/A Build 17763

### Version

20.0.1

### Expected behavior

External IdP integration with ADFS not fail.

### Actual behavior

External IdP integration with ADFS may fail.

### How to Reproduce?

Integrate ADFS as an external IdP using the OIDC identity provider.

**Comment by rmartinc** [comment]
@y-tabata The `relayState` used by the saml provider is a combination of `<state>.<client-id>.<tab-id>` and all of them are encode in Base64Url (so normal chars and `-` and `_`). So in normal usage I don't see when the `+` can be added to the relay state. The only option is if the client does not have an UUID id and then the `clientId` (not encoded is used). Can you paste what is the relay state keycloak is sending to ADFS?

**Comment by y-tabata** [comment]
@rmartinc 
As I described above, I reported about not the `relayState` used by the SAML provider, but the `state` used by the OIDC provider.
The state value in the authorization request from Keycloak is like the following.
```
fNXBt19wC-6XTQ_Pu8Ny3WJw8UEf19isZTyo_CRoqno.uDZnAGhOXWs.9sfeNnXrSVScx9%2BnmfbFfw
```

**Comment by rmartinc** [comment]
HI @y-tabata!

So you are configuring ADFS as OIDC provider, right? (I thought it was as a saml provider, ADFS can be both, but it's more or less the same idea.)

The state is `<state>.<tab-id>.<client-id>` so the `+` is in the client-id. Looking the code [here](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/broker/provider/util/IdentityBrokerState.java#L41-L62), the client-id normally is the the base64url of the UUID of the client ID. But probably in your case it's using the clientId because the ID is not and UUID. Can you please check what is the clientId and ID of the client that initiated the login against keycloak? (The initial one, before clicking in the identity provider to go to the ADFS.)

As commented previously, normally the ID of a client is an UUID and this is used encoded in base64url not containing a `+`sign.

**Comment by y-tabata** [comment]
@rmartinc 
The ClientID is `opscenter`. The ID of the client is unknown.
The phenomenon is, as described above, after version up from 18.0.2 to 20.0.1, external IdP integration with ADFS may fail.

The state value in 18.0.2 is: `o6vRcdpD14Y40V3diGdKua8pAUvsGsxSo1n9H7NXZok.zX3FrrCD6dk.opscenter`.
The state value in 20.0.1 is: `kPj7IIA3IvEA-BTqOuXbCZYENoLjiWBbRMEs-NxR51Y.KswxDCWN_hg.JUDHDiqwSZ%2BvXNblUSl5RA`.

**Comment by rmartinc** [comment]
@y-tabata Checking the history of the file this has been fixed in #15734, so it's fixed in 22.0.2. It was changed to Base64 in #10227 (with `+` and `/`, 22.0.0) and  and to base64url in #15734 (`-` and `_`, 22.0.2). So this is already fixed in keycloak. I'm closing this as duplicate of #15734.

**Comment by y-tabata** [comment]
@rmartinc Thanks for checking!
